### ISSUE TITLE
Artery memory leak

### ISSUE THREAD

**Issue by gakesson** [akka/akka#28390]
We have an application running on Akka 2.5.19 and we use Akka arterty (not cluster) to communicate between two remote actor systems.

We noticed a memory leak in Artery Association when we had an incorrect network routing setup between two actor systems (messages were dropped in one direction), which causes a scala collection in AssociationState to pile up and entries never get released. Eventually the application goes into a continous GC (CMS in our case), until restarted.

When trying to connect to the other actor system below error log is continously printed:

```
Outbound message stream to [akka://AS-A13600@<host>:<port>] failed. Restarting it. Handshake with [akka://AS-A13600@<host>:<port>] did not complete within 20000 ms
akka.remote.artery.OutboundHandshake$HandshakeTimeoutException: Handshake with [akka://AS-A13600@<host>:<port>] did not complete within 20000 ms
```

We have no specific remote artery configuration (except for security), i.e. use the default settings.
This issue is reproducible every time by having a host trying to communicate with a remote actor system:

 - Send 10 messages per second to 3 different remote actors using ActorSelection
 - On remote host, drop network packages using  /sbin/iptables -A INPUT -p tcp --destination-port <port> -j DROP

Issue becomes visible after a few hours on a 384mb heap (260 old generation).
Attaching a snippet on the heapdump taken (3 different Association consuming >99% of retained heap).
![association_leak](https://user-images.githubusercontent.com/4820140/70973480-047a3800-20a6-11ea-9c58-072b42389174.png)

**Comment by patriknw** [comment]
@gakesson Thanks for reporting and describing how to reproduce. Will look into it.

**Comment by patriknw** [comment]
by the way, is it with tcp or aeron-upd transport?

**Comment by gakesson** [comment]
Thanks Patrik!
We're running TCP.

**Comment by patriknw** [comment]
Do you see anything more that could be interesting in the logs? Anything about quarantine? Could you share logs for `akka.remote`?

I haven't been able to reproduce yet. It stays at 1 one instance of AssociationState.

Tried on mac so using pfctl instead of iptables, but should be the same.

**Comment by gakesson** [comment]
It is not AssociationState instances that is leaking but the Promise$DefaultPromise member of the AssociationState.
In the heapdump snippet there is three AssociationState that suffers from leaking - one for each host that drops messages - and each of AssociationState instances contains this promise scala datastructure that is piling up.

If you're unable to reproduce that too I will try to create a stand-alone application for it, replicating the usage of our production application. Please let me know!

**Comment by patriknw** [comment]
ok, I was looking for wrong thing then. I'll try some more tomorrow.

**Comment by hepin1989** [comment]
@gakesson I would be better you share the heap dump with the team.

**Comment by gakesson** [comment]
@hepin1989 I agree but I'm not allowed to do that :-|
If you're unable to reproduce it using the procedure I provided, I will create stand-alone application

**Comment by hepin1989** [comment]
@gakesson Yes , forgot that. A reproducer would be really helpful I think.

**Comment by patriknw** [comment]
I think I understand how this can happen. I haven't been able to reproduce with a "real" scenario so a stand-alone application would be very useful to ensure that we are fixing the right thing.

I think the reason is that the handshake isn't completed, the stream restarted, and that builds up the callbacks on the uniqueRemoteAddressPromise, keeping references to the old incarnations of the streams. It looks like this:

<img width="1174" alt="Screenshot 2019-12-18 at 08 52 45" src="https://user-images.githubusercontent.com/336161/71067147-56898f00-2174-11ea-985e-d29331eeedb5.png">

Does that match your heap dump if you drill down into the DefaultPromise?

**Comment by gakesson** [comment]
Yes, it sure looks as the same.
Attaching a histogram view which shows similar to your's.
![leak_histogram](https://user-images.githubusercontent.com/4820140/71082506-7f6c4d00-2191-11ea-96f5-f337a35c08f7.png)

**Comment by patriknw** [comment]
Good, I have an easy way to reproduce it now, so a stand-alone application might not be needed, but I'd like that you test with a snapshot when I have something ready.

**Comment by gakesson** [comment]
Sure, sounds good but I think the snapshot needs to be based on Akka 2.5.x (we just recently moved to 2.5.27, but 2.6.x will require us to do some code changes).

**Comment by patriknw** [comment]
@gakesson You can try the fix with Akka version `2.5-20191220-123947`. You need to add the snapshot repository to your build, see https://doc.akka.io/docs/akka/current/project/links.html#snapshots-repository

It's on top of latest 2.5.27.

If you want to review the changes you find them in PR https://github.com/akka/akka/pull/28407

**Comment by gakesson** [comment]
@patriknw Sorry for the delay but I've been away on holiday.
I will start this verification today and let you know (it takes a while for the issue to materialize)

**Comment by gakesson** [comment]
@patriknw 
**The correction works for our application**

Test has been running for 24h (previously the issue materialized after ~8h) and there is no build-up in the object retention.

Heap dump also confirms that the memory leak is resolved.

Thanks for the swift help with this issue! May I ask when the next 2.5.x is planned to be released (and if this correction will be incorporated into that release)?

**Comment by patriknw** [comment]
It will be included in upcoming 2.5.28 and 2.6.2 within a few weeks. You should be able to use the timestamped version until then.

**Comment by gakesson** [comment]
Thank you!
